### PR TITLE
feat(engine): FEEL expressions for escalation events

### LINKED ISSUE #10973: Define an escalation code statically or as expression
**Is your feature request related to a problem? Please describe.**
 For an escalation intermediate/end event, I can define the `escalationCode` as a string or dynamically as an expression.

**Describe the solution you'd like** 
Same as in #3792.
 
**Additional context** 
Currently, escalation catch events are not materialized as subscriptions in the state. When an escalation is thrown then the handler walks through the scope hierarchy and checks the escalation codes. In order to be consistent with timer and message events, the `escalation` expression must be evaluated when the event scope is activated.

### PR DESCRIPTION
**Pull request by lzgabel**
## Description

This PR adds FEEL expression support to escalation events.

Supported escalation events:
* end event
* intermediate throw event

## Related issues

closes #10973 

## Definition of Done

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by lzgabel** [comment]
Hi @remcowesterhoud. Please take a moment to review this feature when you have time. 🙇

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: ❌ We shouldn't change the model by setting the `escalationCode` here. We should move the evaluation of this expression to the `BpmnEventPublicationBehavior.throwEscalationEvent` method. This also removes the duplication from this processor and the intermediate throw event processor.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/IntermediateThrowEventProcessor.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/IntermediateThrowEventProcessor.java:213]: ❌ We shouldn't change the model by setting the `escalationCode` here. We should move the evaluation of this expression to the `BpmnEventPublicationBehavior.throwEscalationEvent` method. This also removes the duplication from this processor and the end event processor.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEscalation.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEscalation.java:17]: Here we will run into the same issue which we have with the error event FEEL expressions. I will let you know when we came to a conclusion and then we can apply it to both these PRs.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @lzgabel 👍 As expected here we will run into the same issues we're having with the FEEL expressions for error events. I will let you know when we've reached a decision on that. For now let's put this PR on hold.

I have added some small improvements which you could already work on if you feel like it.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java
Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: ❓ I don't think we need to move it into the `BpmnEventPublicationBehavior.throwEscalationEvent` method, when we fail to evaluate the escalation expression, we should create an incident. 

----

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L139-L146
So I think we can change the `throwEscalationEvent` method signature to this:
```java
public boolean throwEscalationEvent(
      final DirectBuffer throwElementId,
      final DirectBuffer escalationCode,
      final BpmnElementContext context) {
      
      // do something.
      }
```

@remcowesterhoud. WDYT?

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: ❓ I don't think we need to move it into the `BpmnEventPublicationBehavior.throwEscalationEvent` method, and the expression should be evaluated first. If the expression evaluation fails, we will create a incident instead of evaluating the expression after the current element is activated.

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java#L218-L221


@remcowesterhoud. WDYT?

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L139-L146

---

BTW. We can modify the signature of the `throwEscalationEvent` method to avoid changing the model.
```java
public Either<Failure, Boolean> throwEscalationEvent(
      final DirectBuffer throwElementId,
      final DirectBuffer escalationCode,
      final BpmnElementContext context) {

      // do something.
}
```

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: I was thinking we can have the `BpmnEventPublicationBehavior` return an `Either<Failure, Boolean>`. This would involve adding a method here like this which can take care of the evaluation for us:

```java
  public Either<Failure, Boolean> throwEscalationEvent(
      final DirectBuffer throwElementId,
      final ExecutableEscalation escalation,
      final BpmnElementContext context) {
    final var expressionEvaluationResult =
        expressionProcessor.evaluateStringExpressionAsDirectBuffer(
            escalation.getEscalationCodeExpression(), context.getElementInstanceKey());

    if (expressionEvaluationResult.isLeft()) {
      return Either.left(expressionEvaluationResult.getLeft());
    } else {
      final DirectBuffer escalationCode = expressionEvaluationResult.get();
      return Either.right(throwEscalationEvent(throwElementId, escalationCode, context));
    }
  }
```

Note that this calls the current (now private scopes) `throwEscalationEvent()` method. We just need to change the `escalation` parameter to `escalationCode`.

This way we extract this logic from our processors. Making them a bit simpler:

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      final BpmnElementContext activated =
          stateTransitionBehavior.transitionToActivated(activating);
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activated)
          .ifRightOrLeft(
              canBeCompleted -> {
                if (canBeCompleted) {
                  stateTransitionBehavior.completeElement(activated);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activated));
    }
```

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Thanks. @remcowesterhoud. ❤️ 

That's how it was originally planned. I was worried that the current element would be activated multiple times during the resolve incident.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Looking at other parts of the code we do the same thing, so this shouldn't be a problem

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      final var activated = stateTransitionBehavior.transitionToActivated(activating);
      final var completing = stateTransitionBehavior.transitionToCompleting(activated);
      stateTransitionBehavior
          .transitionToCompleted(element, completing)
          .ifRightOrLeft(
              completed -> {
                // terminate all other element instances of the same flow scope
                final var flowScopeContext = stateBehavior.getFlowScopeContext(completed);
                stateTransitionBehavior.terminateChildInstances(flowScopeContext);
              },
              failure -> incidentBehavior.createIncident(failure, completing));
    }
```

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: > Looking at other parts of the code we do the same thing, so this shouldn't be a problem

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/incident/ResolveIncidentProcessor.java#L135-L145

---
```
java.lang.IllegalStateException: Expected to continue processing after incident 2251799813685256 resolved, but failed command not found
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.lambda$attemptToContinueProcessProcessing$1(ResolveIncidentProcessor.java:113) ~[classes/:?]
	at io.camunda.zeebe.util.Either$Left.ifRightOrLeft(Either.java:476) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.attemptToContinueProcessProcessing(ResolveIncidentProcessor.java:100) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.incident.ResolveIncidentProcessor.processRecord(ResolveIncidentProcessor.java:76) ~[classes/:?]
	at io.camunda.zeebe.engine.Engine.process(Engine.java:128) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$3(ProcessingStateMachine.java:264) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:260) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:209) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:185) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[classes/:?]
Caused by: java.lang.IllegalStateException: Expected incident to refer to element in state ELEMENT_ACTIVATING or ELEMENT_COMPLETING, but element is in state ELEMENT_ACTIVATED
```

@remcowesterhoud. Seems like you're missing something? 🤔

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Interesting, I wasn't aware of this. Sorry! I suppose we should wait with transitioning to activated in the processor, which makes sense now that I think about it.

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activating)
          .ifRightOrLeft(
              canBeCompleted -> {
                final BpmnElementContext activated =
                    stateTransitionBehavior.transitionToActivated(activating);
                if (canBeCompleted) {
                  stateTransitionBehavior.completeElement(activated);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));
    }
```

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: By the way, this previous example is not entirely correct. We should use `stateTransitionBehavior.transitionToCompleted(element, completing)` as we do in the other processors.

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: ❓ @remcowesterhoud If we modify it like this:

```java
    @Override
    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
      eventPublicationBehavior
          .throwEscalationEvent(element.getId(), element.getEscalation(), activating)
          .ifRightOrLeft(
              canBeCompleted -> {
                final BpmnElementContext activated =
                    stateTransitionBehavior.transitionToActivated(activating);
                if (canBeCompleted) {
                  final var completing = stateTransitionBehavior.transitionToCompleting(activated);
                  stateTransitionBehavior.completeElement(completing);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));
    }
```

The whole semantic becomes to activate `escalation catch event` first, and then `escalation throw event`, do you think this is as expected?

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Personally I like that. It means the main functionality happens between the `ACTIVATING` and the `ACTIVATED` state of the element instance. However, it does differ from the error end event. So I'll gather some opinions on this and get back to you.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Hi @lzgabel, I have an answer surrounding this topic. Although functionality-wise it does not matter we want to keep it aligned with the error event. As a result we will have to pull the finding of the catch event out of the throw method again. Sorry for the extra work here.

So the processor would look somewhat like this:
- Find the escalation catch event. I suppose this will return an `Either<Failure, Optional<CatchEventTuple>>`
- On success:
    - Transition to activated
    - If catch event found, throw the catch event
    - Transition to completing (if needs to be completed)
    - Transition to completed (if needs to be completed)
- On failure:
    - Create incident
    
I would propose to change the `findEscalationCatchEvent` method to take an `ExecutableEscalation` instead of an `escalationCode`. We can do the expression evaluation in this method so it is still reused by both processors.
The `throwEscalationEvent` will not need the `ExecutableEscalation` anymore and can take a `CatchEventTuple` instead.

I hope this clears things up 🙂 Thanks for keeping me sharp!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEscalation.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEscalation.java:17]: Please have a look at https://github.com/camunda/zeebe/pull/10972#discussion_r1031263828 on how we should approach this for the error event. We should do the same thing here.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Just in case it's not clear, as I may have written this down somewhat confusing. This flow is not all in the `onActivate` method. The transitioning to completed should be part of the `onComplete()` method

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: > The throwEscalationEvent will not need the ExecutableEscalation anymore and can take a CatchEventTuple instead.

https://github.com/camunda/zeebe/blob/4b0d304fb63feac2a4867e56ce9e44e30671e183/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java#L148-L151

❓ Hi @remcowesterhoud. When the catch event is not found, if we do not pass `ExecutableEscalation` here, we will not be able to obtain `escalationCode` at this time. Do you have other ideas?

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: I think we should extract this from this method and pull into a separate method. So in the processor it'd do:

- Catch event is present?
    - Yes? Call the behavior to trigger the catch event and write the escalated record
    - No? Call the behavior to write the not escalated record

I'm a bit stuck on the naming for these 2 methods at the moment. I guess it could be as simple as `throwEscalationEvent(throwElementId, CatchEventTuple, BpmnElementContext)` and `writeNotEscalatedEvent(throwElementId, errorCode, BpmnElementContext)`

Do you think this will work?

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Hi @remcowesterhoud. I think we will still face the problem of evaluating expressions multiple times.

```java

eventPublicationBehavior
          .findEscalationCatchEvent(escalation, activating)
          .ifRightOrLeft(
              catchEventTuple -> {
                final var throwElementId = element.getId();
                if (catchEventTuple.isPresent()) { // catch event is found
                  eventPublicationBehavior.throwEscalationEvent(
                      throwElementId, catchEventTuple.get(), activating);
                } else { // catch event is not found
                  final Optional<DirectBuffer> escalationCodeOptional = element.getEscalation()
                      .getEscalationCode();
                  DirectBuffer escalationCode;
                  if (escalationCodeOptional.isEmpty()) { 
                    final Either<Failure, DirectBuffer> expressionEvaluationResult = expressionProcessor.evaluateStringExpressionAsDirectBuffer(
                        escalation.getEscalationCodeExpression(),
                        activating.getElementInstanceKey());
                    escalationCode = expressionEvaluationResult.get();
                  } else {
                    escalationCode = escalationCodeOptional.get();
                  }
                  eventPublicationBehavior.writeNotEscalatedEvent(
                      throwElementId,
                      escalationCode,
                      activating);
                }
              },
              failure -> incidentBehavior.createIncident(failure, activating));

```

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: We've come full circle 😄 I'll have to accept the we will do the evaluation in the processor and have it duplicated in the other processor. I don't like evaluating it twice so let's avoid this. And since we still want the sequence of the record to be as expected I currently don't see a way around this.
This means the implementation here would be very similar to the error end event 😛

Comment (lzgabel) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: So, let's keep the status quo? Or looking for a different solution? 🤔

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:227]: Yes, let's keep the status quo. Implement it similar to the error end event. We can always improve upon it later if we don't like it.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:724]: 🔧 I know we didn't do this for the other tests either, but I think it would be nice to have the full lifecycle of the intermediate throw event here. So also the activate, activating, activated events.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:758]: 🔧 Same as the previous comment

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:802]: 🔧 Lets also add the events of the throwing and catching events in this assertion.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/incident/EscalationIncidentTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/incident/EscalationIncidentTest.java:30]: ❌ We should also make sure the incidents work as expected when thrown from an intermediate escalation event.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @lzgabel 🍪 Overall it looks good. I have some small improvements I'd like to see 🙂

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationValidator.java
Comment (lzgabel) [reply bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationValidator.java:71]: ❓ @remcowesterhoud. Why does this kind of scene exist? Can you give me an example?

Currently, in the same scope, We should only allow one handler to handle the escalation event. 🤔 

https://github.com/camunda/zeebe/blob/dc0727901b5ccdb891306927fe5f4484a6e0112c/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java#L285-L294

Comment (remcowesterhoud) [reply bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationValidator.java:71]: It is validating the `Escalation`. Multiple catch events can reference the same `Escalation`. Scope is irrelevant here.

image
Screenshot 2022-11-29 at 11 20 09

Comment (remcowesterhoud) [reply bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationValidator.java:71]: This makes me think that it would be nicer to just move this validation to the `EscalationEventDefinitionValidator`. This is where we have this validation for error events as well. This will also make it more clear which element fails.

**Comment by lzgabel** [comment]
Hi @remcowesterhoud. I've fixed all review comments. I'll squash all commits when you're done checking this out.  🙇

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationEventDefinitionValidator.java
Comment (remcowesterhoud) [inline bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationEventDefinitionValidator.java:55]: 🔧 Because we moved this to the `EscalationEventDefinitionValidator` the relevant element will be appended automatically. So we could consider changing the error to `EscalationCode may not be an expression`.

I can also live with this message, just a thought.

**Review by remcowesterhoud — APPROVED** [review]
Great work @lzgabel 🏆 I've added a minor comment, I'll let you decide if you want to change this or not.

Please go ahead and squash the commits!

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationEventDefinitionValidator.java
Comment (lzgabel) [reply bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationEventDefinitionValidator.java:55]: @remcowesterhoud. Let's be consistent with the error catch event. 

> The escalationCode of the escalation catch event is not allowed to be an expression

WDYT?

Comment (remcowesterhoud) [reply bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/EscalationEventDefinitionValidator.java:55]: Fine with me 👍

**Comment by lzgabel** [comment]
> Please go ahead and squash the commits!

Done. 
@remcowesterhoud. Please check this out. ❤️

**Comment by remcowesterhoud** [comment]
Thanks @lzgabel. Please next time split the commits a bit more and add descriptive messages. It makes it easier in the future to see why changes were made. You can refer to the [commit guidelines](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#commit-message-guidelines) for more information about this 🙂 

bors merge

**Pull request merged**
